oauth2client/service_account.py

```diff
@@ -12,96 +12,265 @@
+_PASSWORD_DEFAULT = 'notasecret'
+_AT_LEAST_ONE_KEY = ('Exactly one of private_key_pkcs8_pem and '
+                     'private_key_pkcs12 must be passed to the '
+                     'ServiceAccountCredentials constructor.')
+_PKCS12_ERROR = r"""
+This library only implements PKCS#12 support via the pyOpenSSL library.
+Either install pyOpenSSL, or please convert the .p12 file
+to .pem format:
+    $ cat key.p12 | \
+    >   openssl pkcs12 -nodes -nocerts -passin pass:notasecret | \
+    >   openssl rsa > key.pem
+"""
+_PKCS12_KEY = '_private_key_pkcs12'
+
+
+class ServiceAccountCredentials(AssertionCredentials):
+    """Service Account credential for OAuth 2.0 signed JWT grants.
+
+    Supports
+
+    * JSON keyfile (typically contains a PKCS8 key stored as
+      PEM text)
+    * ``.p12`` key (stores PKCS12 key and certificate)
+
+    Makes an assertion to server using a signed JWT assertion in exchange
+    for an access token.
+
+    This credential does not require a flow to instantiate because it
+    represents a two legged flow, and therefore has all of the required
+    information to generate and refresh its own access tokens.
+
+    At most one of ``private_key_pkcs8_pem`` and ``private_key_pkcs12``
+    can be set.
 
+    Args:
+        service_account_email: string, The email associated with the
+                               service account.
+        private_key_pkcs8_pem: bytes, The content of a PKCS#8 key in PEM
+                               format.
+        private_key_pkcs12: bytes, (Optional) The content of a PKCS#12 key.
+        private_key_password: string, (Optional) Password for PKCS#12 private
+                              key. Defaults to ``notasecret``.
+        scopes: List or string, (Optional) Scopes to use when acquiring
+                an access token.
+        private_key_id: string, (Optional) Private key identifier. Typically
+                        only used with a JSON keyfile. Required for a JSON
+                        keyfile to generate a token assertion.
+        service_account_id: string, (Optional) Service account identifier.
+                            Typically only used with a JSON keyfile. Typically
+                            if the ``service_account_email`` is
+                            ``foo@developer.gserviceaccount.com``, this value
+                            will be ``foo.apps.googleusercontent.com``.
+        user_agent: string, (Optional) User agent to use when sending
+                    request.
+        _token_uri: string, (Optional) The URI to use when requesting an
+                    access token.
+        _revoke_uri: string, (Optional) The URI to use when revoking an
+                     access token.
+        kwargs: dict, Extra key-value pairs to send in the payload body
+                when making an assertion.
 
-class _ServiceAccountCredentials(AssertionCredentials):
-    """Class representing a service account (signed JWT) credential."""
+        Raises:
```

Comment:
Your prose is hard to parse :grinning: 

We discussed two options:
- A single service account credentials class with factory constructors
- Three (or more) service account credentials classes

I preferred the first but said I couldn't justify it with a strong argument. I don't think there is any silver bullet here but think the fewer classes we have, the happier users will be.

We could eliminate the "shoot one-self in foot" problem by removing

- `private_key_pkcs8_pem` (specific to JSON)
- `private_key_pkcs12` (specific to P12)
- `private_key_password` (specific to JSON)
- `private_key_id` (specific to JSON, makes no difference when sent as `kid` in an assertion)
- `service_account_id` (specific to JSON, also **NEVER** used anywhere)

from the constructor and just relying on a `signer` passed to the constructor to encapsulate the relevant info (this was the original goal of the signer's anyhow). This mirrors the way that the `Verifier`/`Signer` classes are implemented.